libsolidity/formal/CHC.cpp

```diff
@@ -315,11 +321,16 @@ void CHC::endVisit(FunctionDefinition const& _function)
 		shouldAnalyze(*m_currentContract)
```

Comment:
Can it handle public library functions too? I.e. will `address(this).balance` be correctly treated as the balance of the contract rather than a separate thing? What about transfers to the contract initiated by the library?

If these are expected to work, can you add some tests that use libraries?

Reply:
Hmm now I'm not sure. `internal` library functions should be the same as free functions, and those should take the context contract as the contract for all that stuff.
I'll check public library functions.

Reply:
Ok you just found a bug in the way we encode `delegatecall` :)
```
library L {
	function l(address payable a) public {
		a.transfer(1);
	}
}

contract C {
	using L for address payable;
	uint x;
	function f(address payable a) public payable {
		require(msg.value > 1);
		uint b1 = address(this).balance;
		a.l();
		uint b2 = address(this).balance;
		assert(b1 == b2);
		assert(x == 0);
	}
}
```

When CHC seeds a `delegatecall` is correctly erases all knowledge about state variables, since we don't know what the called code is doing, that's why the last assertion fails here. However, it incorrectly *does not* erase all balances, so the first assertion holds even though it shouldn't. I'll quickly fix that first in another PR.

Reply:
The fix: https://github.com/ethereum/solidity/pull/11841